### PR TITLE
[IMP] stock,mrp,product: print only barcodes coded in code 128 format

### PR DESCRIPTION
**Pull request by kasp-odoo**
Before this commit
==================
Barcodes are only readable if they are coded in code 128 format.
However, the barcode field currently accepts practically any character.

After this commit
=================
This commit prevents printing the barcode image when generating a barcode not
coded in code 128 format. Instead, the report displays only the barcode's value.

TaskId: 3634187

### CODE REVIEW

File: addons/stock/report/report_lot_barcode.xml [hunk 1]
```diff
@@ -29,12 +29,14 @@
...
-                                        <div t-if="final_barcode" t-att-style="'position:absolute; right:.5px; bottom:.5px'">
+                                        <div t-if="final_barcode and o.valid_code_128" t-att-style="'position:absolute; right:.5px; bottom:.5px'">
                                             <span t-out="final_barcode" t-options="{'widget': 'barcode', 'symbology': 'ECC200DataMatrix', 'img_style': 'width:17mm; height:17mm'}">12345678901</span>
                                         </div>
                                     </t>
                                     <t t-else="">
-                                        <span t-field="o.name" t-options="{'widget': 'barcode', 'img_style': 'width:100%; height:35%'}">Customizable Desk</span>
+                                        <t t-if="o.valid_code_128">
+                                            <span t-field="o.name" t-options="{'widget': 'barcode', 'img_style': 'width:100%; height:35%'}">Customizable Desk</span>
+                                        </t>
```
Comment (kir-odoo) [inline addons/stock/report/report_lot_barcode.xml:39]: I would add `t-if` to the span.

File: odoo/tools/barcode.py [hunk 2]
```diff
@@ -52,3 +52,13 @@ def check_barcode_encoding(barcode, encoding):
...
+
+
+def is_valid_code_128(barcode):
+    """
+    Checks if the given barcode string is valid code 128.
+    :param barcode: The barcode string to check.
+    :return: True if the barcode is valid code 128, False otherwise.
+    :rtype: bool
+    """
+    return bool(re.match(r'^[\x00-\x7F]+$', barcode))
```
Comment (kir-odoo) [inline odoo/tools/barcode.py:64]: need a test for validation of the barcode

File: addons/stock/report/report_stockpicking_operations.xml [hunk 3]
```diff
@@ -145,21 +145,23 @@
...
                                             <t t-if="product_barcode != ml.product_id.barcode">
-                                                <span t-if="ml.product_id and ml.product_id.barcode">
+                                                <span t-if="ml.product_id and ml.product_id.barcode and ml.product_id.valid_code_128">
                                                     <div t-field="ml.product_id.barcode" t-options="{'widget': 'barcode', 'symbology': 'auto', 'width': 400, 'height': 100, 'quiet': 0, 'img_style': 'height:35px;'}">
                                                         <div class="bg-light border-1 rounded d-flex flex-column align-items-center justify-content-center px-1 py-2 opacity-75 text-muted text-center">
                                                             (product barcode)
                                                         </div>
                                                     </div>
                                                 </span>
+                                                <span t-out="ml.product_id.barcode">112233445566</span>
```
Comment (kir-odoo) [inline addons/stock/report/report_stockpicking_operations.xml:164]: Why did you add the Barocode Characters field, per spec you just need to hide the barcode

File: addons/stock/report/report_package_barcode.xml [hunk 4]
```diff
@@ -25,7 +25,7 @@
                             </div>
                         </div>
                     </div>
-                    <div t-else="" class="col-6 text-center">
+                    <div t-elif="o.valid_code_128" class="col-6 text-center">
```
Comment (kir-odoo) [inline addons/stock/report/report_package_barcode.xml:28]: Why don't you check in `if` too? What about the `else` case?

**Review by kir-odoo — CHANGES_REQUESTED** [review]

**Comment by kir-odoo** [comment]
In the commit message no need space between the module name

File: addons/stock/report/report_stockpicking_operations.xml [hunk 5]
```diff
@@ -145,21 +145,23 @@
                                             </div>
                                         </td>
                                         <td class=" text-center h6" t-if="has_serial_number">
-                                            <span t-if="has_serial_number and (ml.lot_id or ml.lot_name)" t-esc="ml.lot_id.name or ml.lot_name" t-options="{'widget': 'barcode', 'humanreadable': 1, 'width': 400, 'height': 100, 'img_style': 'width:100%;height:35px;'}">
+                                            <span t-if="has_serial_number and (ml.lot_id or ml.lot_name) and ml.lot_id.valid_code_128" t-esc="ml.lot_id.name or ml.lot_name" t-options="{'widget': 'barcode', 'width': 400, 'height': 100, 'img_style': 'width:100%;height:35px;'}">
```
Comment (kir-odoo) [inline addons/stock/report/report_stockpicking_operations.xml:148]: Why remove `'humanreadable': 1`

File: addons/product/report/product_product_templates.xml [hunk 6]
```diff
@@ -45,8 +45,8 @@
                     <div class= "text-center o_label_small_barcode">
                         <span class="text-nowrap" t-field="product.default_code"/>
                         <t t-if="barcode">
-                            <div t-out="barcode" style="padding:0" t-options="{'widget': 'barcode', 'symbology': 'auto', 'img_style': barcode_size}"/>
-                            <span class="text-center" t-out="barcode"/>
+                            <div t-if="product.valid_code_128" t-out="barcode" style="padding:0" t-options="{'widget': 'barcode', 'symbology': 'auto', 'img_style': barcode_size}"/>
+                            <div class="text-center" t-out="barcode"/>
```
[the remark this answers is not in the document]
Comment (kasp-odoo) [reply addons/product/report/product_product_templates.xml:49]: Replaced the span tag with the div tag to maintain the proper format between the product's `default_code` and `barcode` text.

File: addons/product/report/product_product_templates.xml [hunk 7]
```diff
@@ -64,14 +64,14 @@
                     <div class="o_label_left_column">
                         <span class="text-nowrap" t-field="product.default_code"/>
                     </div>
-                    <div class="o_label_price_medium text-end">
+                    <div class="text-end" style="padding-top:0;padding-bottom:0">
```
[the remark this answers is not in the document]
Comment (kasp-odoo) [reply addons/product/report/product_product_templates.xml:67]: When the barcode of a product is coded in code 128 format the barcode text is not set as expected so need to make changes in styling.

File: addons/stock/report/report_package_barcode.xml (hunk 4 above)
Comment (kasp-odoo) [reply addons/stock/report/report_package_barcode.xml:28]: Here, if part of the statement just checking for the `valid sscc code` but to validate the `code 128` format have to add a condition and we can't add a condition in the else part of a statement directly so use elif in place of else.

File: addons/stock/report/report_stockpicking_operations.xml (hunk 5 above)
Comment (kasp-odoo) [reply addons/stock/report/report_stockpicking_operations.xml:148]: Added a separate span to print the text of the barcode so removed `humanreadable` attribute.

File: addons/stock/report/report_stockpicking_operations.xml (hunk 3 above)
Comment (kasp-odoo) [reply addons/stock/report/report_stockpicking_operations.xml:164]: Here, if the product's barcode is not coded in Code 128, the barcode image will not be there in the report. However, to improve the user experience, I have added a barcode characters field so that the user can at least see the barcode text in the report.

File: addons/stock/report/report_package_barcode.xml (hunk 4 above)
Comment (kir-odoo) [reply addons/stock/report/report_package_barcode.xml:28]: If valid_sscc is true, it prints the barcode without checking valid_code_128, which might not be correct. Additionally, displaying a blank if both conditions fail is not user-friendly.

Comment (kasp-odoo) [reply addons/stock/report/report_package_barcode.xml:28]: In this case, the barcode is generated from the name of the package and the package name is already printed in the report so I think it will be not user-friendly to print the package name (barcode character) twice in the case of valid_code_128 is false.

**Comment by kir-odoo** [comment]
@kasp-odoo why is the branch for master while the target version is saas-17.2 as defined in task

**Comment by kasp-odoo** [comment]
>@kasp-odoo why is the branch for master while the target version is saas-17.2 as defined in task

@kir-odoo When I started working on this task, saas-17.2 had not been merged, so I created a branch in the master version. By then, saas-17.2 had been merged, so do I need to retarget this PR to saas-17.2, or is it still required in the master?

**Comment by kir-odoo** [comment]
> > @kasp-odoo why is the branch for master while the target version is saas-17.2 as defined in task
> 
> @kir-odoo When I started working on this task, saas-17.2 had not been merged, so I created a branch in the master version. By then, saas-17.2 had been merged, so do I need to retarget this PR to saas-17.2, or is it still required in the master?

Seriously? saas-17.2 was the targeted version when you started the task, and that time even saas-17.3 was about to release. Why was a branch created in master?

Anyway retarget it please.

**Comment by Whenrow** [comment]
I would rather make a change into the barcode widget that create the image. So It will work for future barcode field without the need to create many computed fields `is_valid_barcode`

**Comment by kir-odoo** [comment]
> rather

Hello @Whenrow Okay! I've tried with different generic approach can you please check?

**Comment by Whenrow** [comment]
robodoo r+
thank you

**Comment by duong77476-viindoo** [comment]
A dummy question, this issue has been seen a lot in stable version, can i back port it ?
